static/panes/tool-input-view.js

```diff
@@ -0,0 +1,223 @@
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+'use strict';
+
+var FontScale = require('../fontscale');
+var monaco = require('monaco-editor');
+var _ = require('underscore');
+var $ = require('jquery');
+var ga = require('../analytics');
+var monacoConfig = require('../monaco-config');
+var local = require('../local');
+
+require('../modes/asm-mode');
+
+function ToolInputView(hub, container, state) {
+    state = state || {};
+    this.container = container;
+    this.eventHub = hub.createEventHub();
+    this.domRoot = container.getElement();
+    this.domRoot.html($('#tool-input').html());
+    this.source = state.source || '';
+    this._currentDecorations = [];
+    var root = this.domRoot.find('.monaco-placeholder');
+
+    this.settings = JSON.parse(local.get('settings', '{}'));
+
+    this.editor = monaco.editor.create(root[0], monacoConfig.extendConfig({
+        value: '',
+        language: 'plaintext',
+        readOnly: false,
+        glyphMargin: true,
+    }));
+
+    this._toolId = state.toolId;
+    this._toolName = state.toolName;
+    this._compilerId = state.compilerId;
+    this._editorId = state.editorId;
+
+    this.awaitingInitialResults = false;
+
+    this.initButtons(state);
+    this.initCallbacks();
+
+    this.setTitle();
+    this.onSettingsChange(this.settings);
+    this.eventHub.emit('toolInputViewOpened', this._toolid);
+    ga.proxy('send', {
+        hitType: 'event',
+        eventCategory: 'OpenViewPane',
+        eventAction: 'toolInputView',
+    });
+}
+
+ToolInputView.prototype.initButtons = function (state) {
+    this.fontScale = new FontScale(this.domRoot, state, this.editor);
+
+    this.topBar = this.domRoot.find('.top-bar');
+};
+
+ToolInputView.prototype.initCallbacks = function () {
+    this.fontScale.on('change', _.bind(this.updateState, this));
+
+    this.eventHub.on('compilerClose', this.onCompilerClose, this);
+    this.eventHub.on('toolClosed', this.onToolClose, this);
+    this.eventHub.on('toolInputViewCloseRequest', this.onToolInputViewCloseRequest, this);
+    this.eventHub.on('settingsChange', this.onSettingsChange, this);
+    this.eventHub.on('setToolInput', this.onSetToolInput, this);
+
+    this.container.on('resize', this.resize, this);
+    this.container.on('shown', this.resize, this);
+    this.container.on('destroy', this.close, this);
+
+    this.container.layoutManager.on('initialised', function () {
+        // Once initialized, let everyone know what text we have.
+        this.maybeEmitChange();
+    }, this);
+    this.eventHub.on('initialised', this.maybeEmitChange, this);
+
+    this.editor.getModel().onDidChangeContent(_.bind(function () {
+        this.debouncedEmitChange();
+        this.updateState();
+    }, this));
+
+    this.cursorSelectionThrottledFunction =
```

Comment:
It looks like the selection isn't part of the state (the `getState()` stuff). We should either include the selection and keep this, or continue not to store the selection as part of the URL etc and then all the machinery here about "selection change' can be removed.

Either is fine with me: the selection highlighting is useful if we think it'll be handy for folks to share links with areas of the input highlighted already (ie their last selection). This is handy on the source and ASM views, but it's more questionable on tool inputs.